orttraining/orttraining/python/training/ortmodule/_io.py

```diff
@@ -158,7 +175,7 @@ def _combine_input_buffers_initializers(params, onnx_input_names, input_info, bu
         * Initializers: computed from original PyTorch model parameters
     """
 
-    def _expand_inputs(current_input, non_none_inputs):
+    def _expand_inputs(current_input, non_none_inputs, name=""):
```

Comment:
which scenario the `name` is needed?

Reply:
`name` is needed for the case of kwargs having nested dictionaries. On line 214 of the new code, we need to look up the tensor associated with a name (this name represents the flattened dictionary inside kwargs). It is not possible to look up the name inside kwargs. Which is why, we first collect the flattened kwargs, and then look up the name.

Consider the

```py
kwargs = {
    "batch": {
        "one_value": torch.randn(N, D_in, device=device),
        "two_value": {
            "three_value": torch.randn(N, D_in, device=device),
            "four_value": torch.randn(N, D_in, device=device),
        },
        "five_value": {
            "six_value": torch.randn(N, D_in, device=device),
            "seven_value": {"eight_value": torch.randn(N, D_in, device=device)},
        },
    }
}
```

Now, names in the onnx model inputs will be:
```py
[batch_one_value, batch_two_value, batch_two_value_three_value, batch_two_value_four_value, ...]
```

It would be easier if I flattened the kwargs and looked up the tensors associated with these names inside `flattened_kwargs`.

Reply:
Do you think there will be any problem when we iterate the dict:
"
{
    "three_value": torch.randn(N, D_in, device=device),
    "four_value": torch.randn(N, D_in, device=device),
}
"
the order might be different for different runs?

Reply:
In this particular part of the code, it does not matter. I only flatten the kwargs into another dictionary. The iteration of the inputs happens over the onnx graph inputs (not the kwargs dictionary).

But to answer your question about python dict oredering: pyuthon 3.7+ guarantees that the order of the iteration over a dict is the same as the insertion order.